Scrape Supported Protocols and generate SupportedProtocols.md

sillyfrog opened the pull request:
* Tool to build the document.
* Add reference to new doc.

* Other
  - Move existing python unit tests to Python3
  - Use Python3's pylint
  - Fix pylint issues found after move.
  - Convince CI/Travis to install/use convince Python3 stuff.

Fixes #743

sillyfrog committed:
Initial Scrape Supported Devices script

sillyfrog committed:
Generate SupportedProtocols script and output

sillyfrog committed:
Minor consistency updates

crankyoldgit commented:
Looks good to me. But I'll wait for the _travis_ / lint checks to come back before I click approve.

crankyoldgit committed:
Add Supported protocols to README.md

crankyoldgit reviewed (changes requested):
Sadly, a bunch-o-linter issues.

```
2.26s$ pylint {src,test,tools}/*.py
Using config file /home/travis/build/markszabo/IRremoteESP8266/pylintrc
************* Module scrape_supported_devices
C:  9, 0: Line too long (82/80) (line-too-long)
C: 24, 0: Wrong hanging indentation (add 6 spaces).
    time.asctime()
    ^     | (bad-continuation)
C: 25, 0: Wrong hanging indentation.
)
^     |   | (bad-continuation)
W: 29, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 30, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 31, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 32, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 33, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 37, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 38, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 39, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 40, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W: 41, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 42, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 43, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W: 44, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W: 45, 0: Bad indentation. Found 16 spaces, expected 8 (bad-indentation)
W: 46, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 50, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 51, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 52, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 53, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 54, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W: 55, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 56, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 57, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 58, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 59, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W: 60, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W: 61, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W: 62, 0: Bad indentation. Found 16 spaces, expected 8 (bad-indentation)
W: 63, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 64, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 68, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 69, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 70, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 71, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 72, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 73, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 74, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W: 75, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 76, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 77, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W: 78, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 79, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W: 80, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 81, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 82, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
C: 83, 0: Line too long (83/80) (line-too-long)
W: 83, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W: 84, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 85, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 86, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 87, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 91, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 93, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 94, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 95, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 96, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 97, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 98, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W: 99, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:100, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W:101, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:102, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:106, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:107, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:108, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:114, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:117, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:123, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:129, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:130, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:134, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:135, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:139, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:140, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:141, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:142, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:143, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W:144, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:145, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:146, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W:147, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W:148, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:149, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W:150, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W:151, 0: Bad indentation. Found 16 spaces, expected 8 (bad-indentation)
W:152, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W:153, 0: Bad indentation. Found 16 spaces, expected 8 (bad-indentation)
W:154, 0: Bad indentation. Found 16 spaces, expected 8 (bad-indentation)
W:155, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:159, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:163, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:164, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:165, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:166, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:170, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:171, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:172, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:173, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:174, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W:175, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:176, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:177, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:178, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:179, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:180, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
C:180, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
W:181, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:182, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:183, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:184, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:185, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:186, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:187, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
C:187, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
W:188, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:189, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:190, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:191, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:192, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:194, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:195, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:196, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:198, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:199, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
C:200, 0: Line too long (83/80) (line-too-long)
W:200, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:201, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:203, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:204, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:205, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:206, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:207, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:208, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W:209, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W:210, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W:211, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W:212, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:213, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W:214, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
W:216, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:226, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:227, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:229, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:230, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:232, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:233, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:234, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:235, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
C:235, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
W:236, 0: Bad indentation. Found 8 spaces, expected 4 (bad-indentation)
W:237, 0: Bad indentation. Found 12 spaces, expected 6 (bad-indentation)
C:237, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
W:241, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W:133, 0: Redefining built-in 'exit' (redefined-builtin)
C:  1, 0: Missing module docstring (missing-docstring)
E:  3, 0: Unable to import 'pathlib' (import-error)
C: 28, 0: Missing function docstring (missing-docstring)
C: 36, 0: Missing function docstring (missing-docstring)
C: 38, 8: Variable name "fn" doesn't conform to snake_case naming style (invalid-name)
C: 49, 0: Missing function docstring (missing-docstring)
C: 51, 8: Variable name "fn" doesn't conform to snake_case naming style (invalid-name)
C: 67, 0: Missing function docstring (missing-docstring)
C: 71, 8: Variable name "fn" doesn't conform to snake_case naming style (invalid-name)
C: 90, 0: Argument name "fn" doesn't conform to snake_case naming style (invalid-name)
C:105, 0: Missing function docstring (missing-docstring)
W:106, 4: Global variable 'args' undefined at the module level (global-variable-undefined)
C:106, 4: Constant name "args" doesn't conform to UPPER_CASE naming style (invalid-name)
C:133, 0: Missing function docstring (missing-docstring)
C:138, 0: Argument name "fn" doesn't conform to snake_case naming style (invalid-name)
C:138, 0: Missing function docstring (missing-docstring)
C:158, 0: Argument name "fn" doesn't conform to snake_case naming style (invalid-name)
C:158, 0: Missing function docstring (missing-docstring)
C:162, 0: Missing function docstring (missing-docstring)
C:169, 0: Missing function docstring (missing-docstring)
R:169, 0: Too many local variables (17/15) (too-many-locals)
C:236,12: Variable name "fn" doesn't conform to snake_case naming style (invalid-name)
R:169, 0: Too many branches (13/12) (too-many-branches)
R:169, 0: Too many statements (52/50) (too-many-statements)
-----------------------------------
Your code has been rated at 6.21/10
The command "pylint {src,test,tools}/*.py" exited with 30.
```

crankyoldgit commented:
See: https://github.com/markszabo/IRremoteESP8266/wiki/Library-Maintainers-Guide#python

I strongly suggest installing [yapf](https://github.com/google/yapf) and run it, which should clean up most of those issues.

sillyfrog commented:
Just checking it out now. I normally use `flake8` and `black` for auto formatting. Will update shortly.

crankyoldgit committed:
Fix bulk of linter issues by running `yapf`

FYI @sillyfrog

crankyoldgit commented:
After `yapf` (my prev commit):

```
************* Module scrape_supported_devices
C:  9, 0: Line too long (82/80) (line-too-long)
C:177, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
C:184, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
C:231, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
C:233, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
W:130, 0: Redefining built-in 'exit' (redefined-builtin)
C:  1, 0: Missing module docstring (missing-docstring)
E:  3, 0: Unable to import 'pathlib' (import-error)
C: 26, 0: Missing function docstring (missing-docstring)
C: 34, 0: Missing function docstring (missing-docstring)
C: 36, 6: Variable name "fn" doesn't conform to snake_case naming style (invalid-name)
C: 47, 0: Missing function docstring (missing-docstring)
C: 49, 6: Variable name "fn" doesn't conform to snake_case naming style (invalid-name)
C: 65, 0: Missing function docstring (missing-docstring)
C: 69, 6: Variable name "fn" doesn't conform to snake_case naming style (invalid-name)
C: 88, 0: Argument name "fn" doesn't conform to snake_case naming style (invalid-name)
C:103, 0: Missing function docstring (missing-docstring)
W:104, 2: Global variable 'args' undefined at the module level (global-variable-undefined)
C:104, 2: Constant name "args" doesn't conform to UPPER_CASE naming style (invalid-name)
C:130, 0: Missing function docstring (missing-docstring)
C:135, 0: Argument name "fn" doesn't conform to snake_case naming style (invalid-name)
C:135, 0: Missing function docstring (missing-docstring)
C:155, 0: Argument name "fn" doesn't conform to snake_case naming style (invalid-name)
C:155, 0: Missing function docstring (missing-docstring)
C:159, 0: Missing function docstring (missing-docstring)
C:166, 0: Missing function docstring (missing-docstring)
R:166, 0: Too many local variables (17/15) (too-many-locals)
C:232, 8: Variable name "fn" doesn't conform to snake_case naming style (invalid-name)
R:166, 0: Too many branches (13/12) (too-many-branches)
R:166, 0: Too many statements (52/50) (too-many-statements)

-----------------------------------
Your code has been rated at 9.30/10

```

crankyoldgit commented:
P.S. I'm sorry `pylint` etc is harsh. It does make more readable and perhaps better code.

But yes, I suffer under it's scrutiny too. :-( I just habitually check my code against all the linters & unit tests before I `push` so people don't see my horrible code before it's public. ;-P

sillyfrog committed:
pylint and yapf code, including vscode config

sillyfrog committed:
Merge branch 'master' of github.com:sillyfrog/IRremoteESP8266

sillyfrog commented:
I have just made a bunch of changes that should keep pylint happy - it certainly is harsh!

I have my editor configured to use `flake8` and `black`, so have also committed the config for that so if others edit the project in `vscode`, they'll also have the correct config.

Cheers.

crankyoldgit commented:
> I have just made a bunch of changes that should keep pylint happy - it certainly is harsh!

It certainly **still** is:

```
************* Module scrape_supported_devices
C:195, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
C:202, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
C:249, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
C:251, 0: Unnecessary parens after 'print' keyword (superfluous-parens)
E:  3, 0: Unable to import 'pathlib' (import-error)
R:183, 0: Too many local variables (17/15) (too-many-locals)
R:183, 0: Too many branches (13/12) (too-many-branches)
R:183, 0: Too many statements (52/50) (too-many-statements)
```

I'd suggest suppressing the `pathlib` one. See: https://google.github.io/styleguide/pyguide.html#214-decision on how to suppress. No idea what it's complaining about.

The `print` ones are easy fixes.

The last three should be solved by breaking up the `main` code block. i.e. Move the generation of the output to it's own function/method etc.

> 
> I have my editor configured to use `flake8` and `black`, so have also committed the config for that so if others edit the project in `vscode`, they'll also have the correct config.

Awesome! Thanks!

sillyfrog commented:
Weird re there are still errors, vscode is not showing them, but running from the command line I see them.

Just this with re-factoring:
```
$ pylint tools/scrape_supported_devices.py 
************* Module scrape_supported_devices
tools/scrape_supported_devices.py:193:0: R0913: Too many arguments (6/5) (too-many-arguments)
```
Ha ha ha ha!!!

I'll sort this and then update the PR

sillyfrog commented:
Best I can tell that pylint output is Python 2. Should I just put ignore statements in, or should we try and get the test to actually run with Python 3? (Given Python 2 is end of life in 6 months, we're going to have to do at some point...)

sillyfrog commented:
I pulled the raw Travis log down and saw it was Python 2.

I don't know anything about Travis, but assuming it's running in Docker.

Something like the following should get the test going:
```
apt-get install python3-pip -y
pip3 install pylint
python3 -m pylint {src,test,tools}/*.py
```

This is (currently) giving the following:
```
root@8329877427bd:/ir# python3 -m pylint {src,test,tools}/*.py
************* Module src/*.py
src/*.py:1:0: F0001: No module named src/*.py (fatal)
************* Module test/*.py
test/*.py:1:0: F0001: No module named test/*.py (fatal)
************* Module auto_analyse_raw_data
tools/auto_analyse_raw_data.py:11:0: R0205: Class 'RawIRMessage' inherits from object, can be safely removed from bases in python3 (useless-object-inheritance)
tools/auto_analyse_raw_data.py:65:11: R1716: Simplify chained comparison between the operands (chained-comparison)
tools/auto_analyse_raw_data.py:296:9: R1714: Consider merging these comparisons with "in" to "state not in ('HS', 'BS')" (consider-using-in)
************* Module auto_analyse_raw_data_test
tools/auto_analyse_raw_data_test.py:3:0: E0401: Unable to import 'StringIO' (import-error)
```

crankyoldgit commented:
Yeah. Lets move to Python3's pylint. I'll take care of my code in a min, and push to your repo.

sillyfrog committed:
Really get pylint happy (with Python 3)

sillyfrog commented:
Cool, cheers. I have just pushed the changes to get `pylint` to pass on this new code.

sillyfrog commented:
FYI: this looks like the correct command to run to also cover any potential future changes:
```
python3 -m pylint -d F0001 {src,test,tools}/*.py
```

crankyoldgit committed:
Convert tools/*.py and travis to use Python3
- A number of Python2.7 -> Python3-isms.
- Convert unit tests to run only in Python3.
- Changes in unit test code for Py2.7->3 differences.
- Update travis config to install/use Python3 pylint etc.

FYI @sillyfrog

crankyoldgit commented:
Fingers crossed. I've changed everything to have it now work with Python3 and use it in future.

/me waits for _travis_ to do it's thing ...

crankyoldgit committed:
Attempt to appease the Travis / pip gods.

crankyoldgit committed:
Now with more sudo.

crankyoldgit committed:
Another travis approach for python3

crankyoldgit committed:
Once more

crankyoldgit committed:
Trying a travis matrix approach.

crankyoldgit committed:
Revert "Trying a travis matrix approach."

This reverts commit 5bdd116ae1fe59738cb19f1d38e51900c398afdf.

crankyoldgit committed:
Another try

crankyoldgit committed:
Give up on using pip, and use apt-get instead.

crankyoldgit reviewed (approved):
Yay. It finally passes all the Travis checks!

sillyfrog commented:
Further on this one. Now you have unified things so much, is it worth using the scrape code to generate much of the repeating code in the examples, especially the `IRMQTTServer` (which is awesome, only just started looking at it). It could also generate chunks of `IRremoteESP8266.h` and some of the other `IR*` file sections. It outputs the relevant `.h` files that are then included as required.

I guess it only really matters if there are a lot more new codes coming in, but I feel like that will happen all the time...

crankyoldgit commented:
I've already started "automating" parts of `IRMQTTServer`'s code or moving it to simple/common library functions, just to save code space and effort. So, yes, it is ripe for it. It is also the place I make mistakes of omission the most. i.e. Forgetting to add support for new protocols properly.

Re: `IRremoteESP8266.h`, I can see it doing the `SEND_BLAH` and `DECODE_BLAH` defines, but not the `decode_type_t` structure which has to retain it's order. Given we typically add protocols one at a time, and the library doesn't work properly (unit tests etc) if those are not added, auto generating it is a "high hanging fruit". Sure, do it if you want, I won't stop you, but there are better things we could do IMHO.

Yes, we seem to add a protocol about every 2+ weeks it feels.
I'd automate sanity checks first, or documentation, before generating the source code itself.
e.g. "Does every protocol have a protocol_to_Str() etc?", "Does every protocol have SEND_BLAH, DECODE_BLAH for it?". or "Here are the controllable settings (and function names) for each supported A/C etc", "Here are a list of the TODOs in the code that we need help with or confirmed."
